akka-projection-jdbc/src/main/scala/akka/projection/jdbc/internal/JdbcProjectionImpl.scala

```diff
@@ -29,10 +29,56 @@ import akka.projection.internal.ProjectionSettings
 import akka.projection.internal.RestartBackoffSettings
 import akka.projection.internal.SettingsImpl
 import akka.projection.javadsl
-import akka.projection.jdbc.javadsl.JdbcSession
+import akka.projection.jdbc.JdbcSession
+import akka.projection.jdbc.scaladsl.JdbcHandler
+import akka.projection.scaladsl
+import akka.projection.scaladsl.Handler
 import akka.projection.scaladsl.SourceProvider
 import akka.stream.scaladsl.Source
 
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[projection] object JdbcProjectionImpl {
+
+  private[projection] def createOffsetStore[S <: JdbcSession](sessionFactory: () => S)(
+      implicit system: ActorSystem[_]) =
+    new JdbcOffsetStore[S](JdbcSettings(system), sessionFactory)
+
+  private[projection] def adaptedHandlerForExactlyOnce[Offset, Envelope, S <: JdbcSession](
+      projectionId: ProjectionId,
+      sourceProvider: SourceProvider[Offset, Envelope],
+      sessionFactory: () => S,
+      handler: JdbcHandler[Envelope, S],
+      offsetStore: JdbcOffsetStore[S])(implicit system: ActorSystem[_]) = {
+
+    new Handler[Envelope] {
+
+      private val logger = Logging(system.classicSystem, classOf[JdbcProjectionImpl[_, _, _]])
+
+      override def process(envelope: Envelope): Future[Done] = {
+        val offset = sourceProvider.extractOffset(envelope)
+        // this scope ensures that the blocking DB dispatcher is used solely for DB operations
+        implicit val executionContext: ExecutionContext = offsetStore.executionContext
+        JdbcSession
+          .withSession(sessionFactory) { sess =>
```

Comment:
I wonder if the connection object should just be passed instead of, or in addition to the session object. I see that it's already available in the session as `Session.conn`, but it seems like you're encouraging users to use `withConnection` in the tests.  This seems like extra and maybe unnecessary hoops to jump through to access the connection since `withConnection` doesn't do anything special anyway.

Reply:
It's not always possible to access the connection. Hibernate for instance, doesn't provide an accessor. If you want to have access to a connection you have to provide a lambda to Hibernate. Hence the choice of copying that pattern.

I think this will become more clear when we have a full Hibernate example in the docs. 

True that if you are using pure JDBC and have access to the connection object, you could use it directly. 

Unfortunately, we have to code for the most restrictive API (Hibernate), otherwise we won't be able to support it.